File: Python/BinaryTree.py
from functools import *


class Node:

    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None
        self.parent = None

    def __repr__(self):
        return hex(id(self))

    def __str__(self):
        return f"( { self.__repr__() } : { self.data } )"


def BST(object):

    def __init__(self, root=Node()):
        self.root = root

    def insert(self, node):
        self._insert(node, self.root)


    def _printTree(self):
        self.printTree()

    def _insert(self, node, cur_node):

        if node.data < cur_node.data:
            if cur_node.data.left == None:
                cur_node.left = node
            else:
                self._insert(node, cur_node.left)

        else:
            if cur_node.data.right == None:
                cur_node.right = node
            else:
                self._insert(node, cur_node.right)

    def printTree(self):
        pass

if __name__ == "__main__":
    b = BST(Node(15))
    b.insert(Node(10))
    b.insert(Node(20))
    print( b.root.data )
    print( b.root.left.data )
    print( b.root.right.data )

File: C++/BinaryTree.cpp
#include "iostream"
#include "vector"


template<class T>
class Node{

	public:
		T	  data;
		Node* left  = NULL;
		Node* right = NULL;
}



template<class T>
class BinaryTree{
	public:
		Node<T>* root = NULL;

		BianryTree() {};
		~BinaryTree() {};

		void AddNode(Node<T>* node, Node<T>* curNode);
		Node<T>* Search(T data);
		Node<T>* Delete(T data);

}

template<class T>
Node<T>* BinaryTree<T>::Search(T data){
	Node<T>* tempNode = this->root;
	
	while (1){
		
		if (data == tempNode->data || (tempNode->left == NULL && tempNode->right == NULL)){
			return tempNode
		}

		if ( tempNode->left->data < data ){
			if (tempNode->left == NULL){
				tempNode = tempNode->right;
			}
			else if(tempNode->right == NULL){
			tempNode = tempNode->left;
			}
		}
		else{
			if (tempNode->right == NULL){
				tempNode = tempNode->left;
			} 
			else if (tempNode->right == NULL){
				tempNode = tempNode->right;
			}
		}
	}
}

template<class T>
void AddNode(Node<T>* node, Node<T>* curNode ){

}




